atomix/cluster/src/main/java/io/atomix/cluster/messaging/impl/DefaultClusterCommunicationService.java

```diff
@@ -273,11 +285,15 @@ public boolean isRunning() {
 
     @Override
     public CompletableFuture<byte[]> apply(final Address sender, final byte[] bytes) {
+      if (blockingAddresses.contains(sender)) {
```

Comment:
My only problem is that this is now in the production code. But I don't have a better idea. :woman_shrugging:

Reply:
Yeah thought the same, but had no better idea. Some can argue that no one use it, just us. So we total free about the interfaces :D but in general I agree.

Reply:
Is the question how weird it really is to have such methods in the API, in general I think it makes sense to have the possibility to block some senders, like blacklisting. So it is traffic/communication related, fits to the interface kind of. But yeah since I was so unsure I opened the PR :cry: :laughing:

Reply:
Isn't `ClusterCommunicationService` an interface? It would be a matter of injecting the correct service, no? Or maybe it has to be lower at the messaging service, not sure. But I agree with Deepthi, if possible I would prefer to have test purpose code in test code.

Reply:
Problem is this is needed in integration tests, so we don't have a way to inject services yet.

Reply:
Exactly. In the integration tests we are not able to exchange the implementations, not on the fly nor on start of the brokers, since everything is bundled inside the broker as you know :) Which I think makes sense. 

Alternative would be to do it with containers and try to partition them

Reply:
Other alternative we cast it in the IT tests: `((DefaultClusterCommunicationService) atomix.getCommunicationService())`, which mega not nice, since in IT tests we shouldn't rely on such internals, but then we would need it on the API. :shrug: 

I don't know. If no one else has a better option I would go with it for now. It is at least no public api.

Reply:
So we have four options, correct? 

1. Use the above proposal, modifying the production API for testability since we can't swap out implementations during tests. This has the downside of coupling high level tests to implementation details.
1. Refactoring our codebase to allow injecting a different implementation in integration tests. Generally though, even if we could swap the implementation at runtime, the communication layer is very much implementation details, so having the ability to directly modify the instance in the QA tests is a code smell and couples our tests to the implementation.
1. Use container based tests and `tc` on the nodes to simulate an asymmetric partition. Toxiproxy isn't really useful since it only lets you stop a proxy completely, but doesn't let you block requests only for one node. Similarly, using Docker network commands, you can only disconnect a container from the network, which wouldn't help in this case.

Reply:
Any thoughts @npepinpe

Reply:
I would lean towards a proxy based and/or container based solution, since that's the only way to really write tests with asymmetric partitions without coupling them to the implementation. But it makes me think, do we need to write the tests at such a high level? There's a lower level of integration where we have access to the communication service implementation and maybe it's sufficient to test there? I guess it depends on the tests.

Reply:
Yes if we would go the container option then we could reuse logic we have in our chaos experiments https://github.com/zeebe-io/zeebe-chaos/blob/master/chaos-workers/chaos-experiments/scripts/disconnect-leaders.sh#L49 

But I think this would make the tests slower, which is why I was not sure. Should I try it out?

Reply:
> I would lean towards a proxy based and/or container based solution, since that's the only way to really write tests with asymmetric partitions without coupling them to the implementation. But it makes me think, do we need to write the tests at such a high level? There's a lower level of integration where we have access to the communication service implementation and maybe it's sufficient to test there? I guess it depends on the tests. I guess I would lean towards narrowing the scope of integration - that has the benefit that you control the components better, and the blast radius is smaller leading to failures being easier to diagnose.

Agree for certain tests it might make sense, but in this case we need everything around. Clients, deployment distribution etc.

Reply:
Just a thought here: Could we also achieve the same result by modifying iptables rules inside containers? Similar to this: https://stackoverflow.com/a/30467578

Reply:
Hey Ole this is exactly what we do in our chaos experiments https://github.com/zeebe-io/zeebe-chaos/blob/master/chaos-workers/chaos-experiments/scripts/disconnect-leaders.sh

Reply:
I find `tc` easier to use than iptables, but YMMV :man_shrugging:

Reply:
Ok sorry @oleschoenburg I think I misunderstood your message. You meant `iptables` not `ip route` sorry :see_no_evil: 

Do you know how to use it for disabling certain ports? I have currently an issue with the testcontainers because they all run on localhost. Via the java api I'm not able to get the container ip (it just returns `localhost`). what I need to block the route is the ip address like `172.25.X.X`.

If not, I might need to first query the ip address via ip route on the other container to block it on the different container.

Reply:
The route command should work with hostnames - would that be enough? :thinking: Let me know if you need anything, brainstorming, pairing, etc., happy to help

Reply:
No it only accepts IP's. https://apple.stackexchange.com/a/332183 I ran into the next issue

Reply:
OK NOW I was able to get it running. I was able to set the capabilities and create the ip route. The test fails because the deployment wasn't be distributed to the partition, which proves that the network partition works !!! 

Please take a look at the new test with TestContainers, happy to get your feedback
https://github.com/camunda-cloud/zeebe/pull/8356/commits/9b25f1bf5d3bfd2a5c769e0e42dfd7d72dfb4268 
@npepinpe @oleschoenburg

Reply:
Sounds like you should also get the real ip by doing this: `container.getContainerInfo().getNetworkSettings().getIpAddress()` 
via https://github.com/testcontainers/testcontainers-java/issues/452#issuecomment-954296099

Reply:
Thanks @oleschoenburg I had a look at it with the debugger. The suggested method returns empty unfortunately, but based on this I found another way.

![ip](https://user-images.githubusercontent.com/2758593/145836936-b0540f44-35f9-48ef-9a4e-0cf4e5db33f6.png)


Seems to work:
```
leaderOneNode.getCurrentContainerInfo().getNetworkSettings().getNetworks().values().stream().findFirst().orElseThrow().getIpAddress()
```

Reply:
One other option which works (but I'm still not sure whether I prefer it or not) is to use multiple networks.

Take two networks, blue and red. Color one node red, one node blue, and one both. This simulates a situation where only one node can speak to the others. It's not quite an asymmetric partition - where A can talk to B, but B cannot talk to A, though, so it's not quite as powerful, but it doesn't rely on any tools other than Docker. 

Just a thought.